#12365 Fix HTTP server support of UNIX filesystem sockets

p12tic opened the pull request:
## Scope and purpose

Fixes #12365 
Fixes #12372 

https://github.com/twisted/twisted/pull/12312 changed HTTPChannel to call setTcpNoDelay() on any transport implementing ITCPTransport. Transports via UNIX socket also implement this interface. As a result, setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, ...) was called on UNIX socket, which fails via:

builtins.OSError: [Errno 95] Operation not supported

It is reasonable to solve the problem simply by ignoring OSError related to TCP NODELAY, because NODELAY is relatively small performance optimization. In almost all cases the caller will not be able to do anything to remedy the situation and will simply continue.

The PR does not include additional tests for HTTPChannel, however tests for UNIX transport and its handling of TCP_NODELAY and TCP_KEEPALIVE flags have been added.

itamarst commented:
Ugggh why is Unix socket a TCP socket :(

adiroiban commented on src/twisted/internet/tcp.py:
```diff
@@ -334,10 +334,16 @@ def logPrefix(self):
 
     def getTcpNoDelay(self):
-        return bool(self.socket.getsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY))
+        try:
```
I think the TCP transport implementation should be left as it is

Many thanks for the report.

We have the [ITCPTransport](https://docs.twisted.org/en/latest/api/twisted.internet.interfaces.ITCPTransport.html) interface that defined the setTcpNoDelay

I think that in the HTTP code we need to conditionally call `self.transport.setTcpNoDelay(True)` only if the transport implements the `ITCPTransport` interface

itamarst replied:
It already does, pretty sure, the issue is that Unix transport pretends to be TCP transport?

adiroiban replied:
True. This comment was before doing more troubleshooting

adiroiban commented:
Thanks Povilas for the PR.

I left a comment inline.

I don't think that we should call TCP specific methods on an Unix trasport.

adiroiban commented:
So the code at line 2336 is this

https://github.com/twisted/twisted/blob/a6c0e2dd1df7f7610aa4efe64a442db829954e44/src/twisted/web/http.py#L2334-L2337


we need to see why the Unix transport is recognized as a TCP transport

adiroiban commented:
There is someting strange here.

To manualy reproduce

```
$ twist web --listen='unix:/tmp/twisted-web:mode=660'
```

In another terminal trigger a new connection

```
$ echo 'not important' |  socat - UNIX-CONNECT:/tmp/twisted-web
```

adiroiban commented:
More troubleshooting see where we end up with TCP interface for the Unix transport

```
(Pdb) self.transport
<_GenericHTTPChannelProtocol #0 on None>
(Pdb) from zope.interface.declarations import providedBy
(Pdb) ca = providedBy(self.transport)
(Pdb) ca._implied
{classImplements(Server, IUNIXTransport): (), <InterfaceClass twisted.internet.interfaces.IUNIXTransport>: (), classImplements(_SendmsgMixin): (), classImplements(Server): (), classImplements(ServerMixin): (), classImplements(Connection, ITLSTransport, ITCPTransport, ISystemHandle): (), <InterfaceClass twisted.internet.interfaces.ITLSTransport>: (), <InterfaceClass twisted.internet.interfaces.ITCPTransport>: (), <InterfaceClass twisted.internet.interfaces.ISystemHandle>: (), classImplements(ConnectionMixin): (), classImplements(FileDescriptor, IPushProducer, IReadWriteDescriptor, IConsumer, ITransport, IHalfCloseableDescriptor): (), <InterfaceClass twisted.internet.interfaces.IPushProducer>: (), <InterfaceClass twisted.internet.interfaces.IProducer>: (), <InterfaceClass twisted.internet.interfaces.IReadWriteDescriptor>: (), <InterfaceClass twisted.internet.interfaces.IReadDescriptor>: (), <InterfaceClass twisted.internet.interfaces.IWriteDescriptor>: (), <InterfaceClass twisted.internet.interfaces.IFileDescriptor>: (), <InterfaceClass twisted.internet.interfaces.IConsumer>: (), <InterfaceClass twisted.internet.interfaces.ITransport>: (), <InterfaceClass twisted.internet.interfaces.IHalfCloseableDescriptor>: (), classImplements(_ConsumerMixin): (), classImplements(_LogOwner, ILoggingContext): (), <InterfaceClass twisted.internet.interfaces.ILoggingContext>: (), classImplements(_SocketCloser): (), classImplements(_AbortingMixin): (), classImplements(object): (), <InterfaceClass zope.interface.Interface>: ()}
```

adiroiban commented:
I think that the issue is here

`t.internet.unix.Server` inherits from `t.internet.tcp.Server`

https://github.com/twisted/twisted/blob/a6c0e2dd1df7f7610aa4efe64a442db829954e44/src/twisted/internet/unix.py#L241-L245


 which inhertis from `twisted.interent.tcp.Connection`

https://github.com/twisted/twisted/blob/a6c0e2dd1df7f7610aa4efe64a442db829954e44/src/twisted/internet/tcp.py#L769-L773

which implements

https://github.com/twisted/twisted/blob/a6c0e2dd1df7f7610aa4efe64a442db829954e44/src/twisted/internet/tcp.py#L209-L212

itamarst commented:
So maybe better to just force the Unix transport to not implement that interface?

p12tic commented:
I'm not sure if making Unix sockets too different from TCP sockets is the right approach. This would be different than the abstractions encoded in the underlying POSIX API. There TCP and Unix sockets are very similar. For example, compare the socket creation:

```
int tcp_socket = socket(AF_INET, SOCK_STREAM, 0);
int unix_socket = socket(AF_UNIX, SOCK_STREAM, 0);
```

Unix sockets are treated effectively as just another addressing mode, just like ipv6 (ipv6 socket would be created via `socket(AF_INET6, SOCK_STREAM, 0);`.

The similarity carries over to UDP and datagram UNIX sockets - one just changes SOCK_STREAM to SOCK_DGRAM.

Furthermore, the similarity is enforced by the fact that SO_KEEPALIVE is supported for UNIX sockets as well.

Therefore I think it makes sense to keep classes related to UNIX sockets as a subclasses of classes related to TCP sockets.

Putting abstractions aside, note that setting TCP_NODELAY may fail anyways (see e.g. https://github.com/thespianpy/Thespian/issues/70 or https://github.com/envoyproxy/envoy/issues/1446). so we may end up needing to do `try ... except` anyway. That would result in abstraction changes being not needed.

itamarst commented:
The possibility for TCP_NODELAY failing makes this a lot more compelling of a workaround, yeah.

adiroiban commented:
One quick and dirty fix is to implement setTcpNoDelay as NOOP for Unix server

glyph commented:
I am inclined to let this slide as an initial fix, given that we shipped something pretty catastrophically broken for a legit use-case, and we should address that before moving on to less significant architectural concerns, but I really don't like the fact that an error occurs here and we just throw it away completely, with no opportunity for the application to recover.

UNIX transports definitely should *not* be implementing an interface called `ITCPTransport`, either. If there are some features we need from it, we should try to address those with a discrete interface.

glyph commented on src/twisted/internet/test/test_unix.py:
```diff
@@ -1014,3 +1016,104 @@ def connect(self, reactor, factory):
+
+    path = "unixpath"
+
+    def test_closePortInProtocolFactory(self):
```
New tests should really have type annotations.
```suggestion
    def test_closePortInProtocolFactory(self) -> None:
```

p12tic commented:
Coverage does not like lines that were changed due to addressing of a pre-existing typing issue. I think it makes sense to ignore this failure for now, because it is completely unrelated to what the PR attempted to fix originally.

p12tic commented:
@glyph @adiroiban @itamarst I think I've addressed your concerns

adiroiban commented on src/twisted/test/test_tcp.py:
```diff
@@ -118,9 +119,9 @@ class MyProtocolFactoryMixin:
-    protocol: Optional[Callable[[], Protocol]] = None
+    protocolConnectionMade: Optional[defer.Deferred[AccumulatingProtocol]] = None
+    protocolConnectionLost: Optional[defer.Deferred[AccumulatingProtocol]] = None
+    createdProtocol: Optional[AccumulatingProtocol] = None
```
Can you please also  update the docstring at line 104

Maybe this can be called something light this ... or `latestProtocol` or `recentProtocol`

```suggestion
    lastProtocol: Optional[AccumulatingProtocol] = None
```

p12tic replied:
Fixed now.

adiroiban commented on src/twisted/internet/tcp.py:
```diff
@@ -337,7 +337,18 @@ def getTcpNoDelay(self):
     def setTcpNoDelay(self, enabled):
-        self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, enabled)
+        try:
+            # There are bug reports about failures when setting TCP_NODELAY under certain conditions
```
I am ok to have this fix here, but this looks like a different fix to the one described in the original bug report.

I think that we need to have a separate newsfragment file for this bugfix.

p12tic replied:
Created a bug report and added newsfragment. Thanks.

adiroiban commented on src/twisted/internet/tcp.py:
```diff
@@ -337,7 +337,18 @@ def getTcpNoDelay(self):
+            # continue.
+            self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, enabled)
+        except OSError:  # pragma: no cover
+            pass
```
Can we conditionally ignore the error only on macOS ?

I don't know under which conditions `TCP_NODELAY` might fail.

For a production code, I prefer to know that the failed.

Maybe emit a log message ?

p12tic replied:
> Can we conditionally ignore the error only on macOS ?

I think we should at least not fail. This would be endless source of frustration. Users will not be amused once they find that they spent hours tracking down rare failure (that may not have proper logs, as seen in the related Django bug), only to find that the cause is an optional optimization that twisted enabled.

Logging the failure is good idea though, thanks. I've implemented that solution.

adiroiban commented on src/twisted/test/test_tcp.py:
```diff
@@ -814,7 +817,7 @@ def _check1cleanup(results, cf1):
     def buildProtocol(self, address):
         self.address = address
-        self.protocol = AccumulatingProtocol()
+        self.createdProtocol = AccumulatingProtocol()
```
I think that we should just delete `MyOtherClientFactory` 

Is this code really used in the tests ?

p12tic replied:
Fixed now.

adiroiban commented on src/twisted/test/test_tcp.py:
```diff
@@ -82,6 +82,7 @@ def cleanUp(self):
         Clean-up for tests to wait for the port to stop listening.
         """
         if self._cleanerUpper is None:
```
Maybe add a comment

```suggestion
        if self._cleanerUpper is None:  # pragma: no cover
           # Under normal test operation, the port should be already stopped.
           # Stop it here in the case the test execution failed without
           # stopping the port.
```

p12tic replied:
Sure, applied, thanks!

adiroiban commented on src/twisted/internet/unix.py:
```diff
@@ -294,6 +294,13 @@ def getHost(self):
     def getPeer(self):
         return address.UNIXAddress(self.hostname or None)
 
+    def getTcpNoDelay(self):
```
I think that we need a significan comment here to exlain why do we have to care about TCP things on an Unix socket

I also think that we should looks at reimplementing the inheritance for `t.i.unix.Server` so that it doesn't inherit from `tcp.Server`

It can inherit from some sort of mixin socket server that doesn't implement any interface.

```suggestion
    def getTcpNoDelay(self):
        """
        FIXME:https://github.com/twisted/twisted/issues/12369
        
        L{twisted.internet.unix.Server} inherits from L{twisted.internet.tcp.Server} which has this method implemented for TCP.
        For Unix socket, this is just a NOOP to avoid
        errors for the code that calls TCP specicific methods thinking that the Unix transport is a TCP transport.
        
```

p12tic replied:
Added, thanks!

adiroiban commented on src/twisted/internet/unix.py:
```diff
@@ -294,6 +294,13 @@ def getHost(self):
+        return False
+
+    def setTcpNoDelay(self, enabled):
+        # This is not supported on UNIX sockets and therefore silently ignored.
```
I tnink that we should reference the followup ticket

This is kind of a workaround for our bad implementantation of `twisted.internet.unix.Server` which should not declare that it implements the ITCPTransport inteface

```suggestion
        """        FIXME:https://github.com/twisted/twisted/issues/12369
        
        This is not supported on UNIX sockets and therefore silently ignored.
        """
```

p12tic replied:
Fixed, thanks!

adiroiban commented on src/twisted/internet/test/test_unix.py:
```diff
@@ -1014,3 +1017,121 @@ def connect(self, reactor, factory):
+
+        return self._connectedClientAndServerTest(check)
+
+    def testFailing(self) -> Deferred[Any]:
```
Can we have a docstring for this test ?

```suggestion
    def test_Failing(self) -> Deferred[Any]:
```

I think that you can also write `async def` tests and yield a deferred and trial should just work 

it should make the tests easier to read

p12tic replied:
Fixed.

p12tic committed:
Do not override .protocol with incompatible class in tcp tests

p12tic committed:
Remove unused code in TCP tests

p12tic committed:
Add better typing to TCP tests

p12tic committed:
Add default arguments to IReactorUNIX interface for better typing

p12tic committed:
Fix HTTP server support of UNIX filesystem sockets

https://github.com/twisted/twisted/pull/12312 changed HTTPChannel to
call setTcpNoDelay() on any transport implementing ITCPTransport.
Transports via UNIX socket also implement this interface. As a result,
setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, ...) was called on
UNIX socket, which fails via:

builtins.OSError: [Errno 95] Operation not supported

To avoid the problem, unix.Server and unix.Client now implement
setTcpNoDelay() and friends that do nothing.

p12tic committed:
Eat OSError when calling setsockopt with TCP_NODELAY

There are bug reports about failures when setting TCP_NODELAY under
certain conditions on macOS:
https://github.com/thespianpy/Thespian/issues/70,
https://github.com/envoyproxy/envoy/issues/1446.

It is reasonable to simply eat errors coming from setting TCP_NODELAY
because TCP_NODELAY is relatively small performance optimization. In
almost all cases the caller will not be able to do anything to remedy
the situation and will simply continue.

p12tic committed:
Add TCP test failure cleanup code path to coverage exception

adiroiban commented on src/twisted/newsfragments/12372.bugfix:
```diff
@@ -0,0 +1 @@
+Fix potential failure of setting TCP_NODELAY on macOS.
```
Thanks for creating the separate ticket.

I have updated to the standard format for the release notes

```suggestion
twisted.internet.tcp.Connection.setTcpNoDelay was updated to fix a potential failure when setting TCP_NODELAY on macOS.
```

p12tic replied:
Fixed, thanks.

adiroiban reviewed (approved):
Many thanks for the updates.
They look good.

I was expecting to have an automated tests in which we listen with an HTTP server over a real Unix socket.

... the title of this PR is "Fix HTTP server support..." but there are no testsfor Unix sockets with an HTTP server

But this PR is already a good start for end to end testing of Unix sockets... so I am not blocking the merge.

adiroiban committed:
Merge branch 'trunk' into 12365-http-fix-unix-socket

adiroiban commented:
Thanks for the updates.

I have enabled auto-merge for this PR.

This is not the ideal fix... but it's a fix.

We have the followup https://github.com/twisted/twisted/issues/12369 if anyone has the energy to untagle the inheritance code for Unix socket transport

p12tic commented:
@adiroiban I think this fix may qualify for separate release as 24.10.1. Right now many people are wasting their time by accidentally installing newest Twisted and then figuring out that they need older version. Probably most Twisted production deployments as HTTP server are affected - in production a reverse proxy will be used and it is recommended it to use unix connection when connecting to nginx.

I can do a PR to stable branch or wherever you want.
